junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala

```diff
@@ -47,35 +55,131 @@ abstract class JUnitTest {
+    lazy val verbose: Boolean = args.contains("-v")
+    lazy val noColor: Boolean = args.contains("-n")
+    lazy val quiet: Boolean = args.contains("-q")
+    lazy val decodeScalaNames: Boolean = args.contains("-s")
+    lazy val logExceptionClass: Boolean = !args.contains("-c")
+    lazy val logAssert: Boolean = args.contains("-a")
 
-    lazy val verbose: Boolean = frameworkArgs.contains("-v")
-    lazy val noColor: Boolean = frameworkArgs.contains("-n")
-    lazy val quiet: Boolean = frameworkArgs.contains("-q")
-    lazy val decodeScalaNames: Boolean = frameworkArgs.contains("-s")
-    lazy val notLogExceptionClass: Boolean = frameworkArgs.contains("-c")
-    lazy val logAssert: Boolean = frameworkArgs.contains("-a")
+    override def toString: String = args.mkString("[", ", ", "]")
+  }
 
-    def formattedTestClass: String = {
-      val (pack, cls) =
-        suiteUnderTestName.splitAt(suiteUnderTestName.lastIndexOf('.'))
-      pack + '.' + yellow(cls.tail)
+  protected final class OutputBuilder private (val argInfo: ArgInfo,
+      total: Int, ignored: Int, failed: Int, output: List[Output]) {
+    import argInfo._
+
+    private[JUnitTest] def this(argInfo: ArgInfo) = this(argInfo, 0, 0, 0, Nil)
+
+    // Builder methods.
+
+    def success(testName: String): OutputBuilder = append(1, 0, 0)(
+        testStartedOutput(testName),
+        testFinishedOutput(testName),
+        successEvent
+    )
+
+    def ignored(testName: String): OutputBuilder = append(0, 1, 0)(
+        testIgnoredOutput(testName),
+        skippedEvent
+    )
+
+    def ignoredClass(): OutputBuilder = append(0, 1, 0)(
+        testIgnoredClassOutput,
+        skippedEvent
+    )
+
+    def assumptionViolated(testName: String) = append(1, 0, 0)(
+        testStartedOutput(testName),
+        testAssumptionViolatedOutput(testName),
+        skippedEvent,
+        testFinishedOutput(testName)
+    )
+
+    def exception(testName: String, msg: String, clazz: String): OutputBuilder = {
+      append(1, 0, 1)(
+          testStartedOutput(testName),
+          testExceptionMsgOutput(testName, msg, clazz),
+          failureEvent,
+          testFinishedOutput(testName)
+      )
     }
 
-    def formattedAssumptionViolatedException: String =
-      "org.junit.internal." + red("AssumptionViolatedException")
+    def assertion(testName: String, message: String): OutputBuilder = {
+      append(1, 0, 1)(
+          testStartedOutput(testName),
+          testAssertionErrorMsgOutput(testName, message),
+          failureEvent,
+          testFinishedOutput(testName)
+      )
+    }
+
+    private def append(t: Int, i: Int, f: Int)(out: Output*) = {
+      new OutputBuilder(argInfo, total + t, ignored + i, failed + f, output ++ out)
+    }
+
+    // Test method.
+
+    private[JUnitTest] def checkOutput(actual: List[Output]): Unit = {
+      val expected = (
+          testRunStartedOutput :: output :::
+          List(testRunFinishedOutput(total, ignored, failed), done)
```

Comment:
Consider using `::`s all the way, instead of using the `List()` constructor for part of it. Maybe format it as follows?
```scala
      val expected = (
          testRunStartedOutput ::
          output :::
          testRunFinishedOutput(total, ignored, failed) ::
          done ::
          Nil
      )
```